libretroshare/src/jsonapi/jsonapi.cpp

```diff
@@ -127,18 +129,16 @@ JsonApiServer::corsOptionsHeaders =
 bool RsJsonApi::parseToken(
         const std::string& clear_token, std::string& user,std::string& passwd )
 {
-	uint32_t colonIndex = 0;
-	uint32_t colonCounter = 0;
+	auto colonIndex = std::string::npos;
 	const auto tkLen = clear_token.length();
 
-	for(uint32_t i=0; i < tkLen && colonCounter < 2; ++i)
-		if(clear_token[i] == ':') { ++colonCounter; colonIndex = i; }
-		else if(!librs::util::is_alphanumeric(clear_token[i])) return false;
+	for(uint32_t i=0; i < tkLen; ++i)
```

Comment:
This is ambiguous. What happens if the token is "user:passwd:passwd" ? then the user will be "user:passwd". That means a passwd cannot contain ":".  In summary, I think that the passwd should be required to be alphanumeric. Otherwise you need to send it in base64 in which case the passwd can be anything and the parsing can be made non ambiguous by ensuiring that the token contains exactly one ":" and that the encoded passwd is what remains after ":".

Reply:
If the token is like "somethingA:somethinchB:somethingC" then user is "somethingA" and password is "somethinchB:somethingC"

Reply:
In fact the token is transferred as a base64 encoded string, RFC7617 has specific rules, IMHO the most relevant parts are:

- For the user-id, recipients MUST support all characters defined in the "UsernameCasePreserved" profile defined in Section 3.3 of [RFC7613], with the exception of the colon (":") character.

- For the password, recipients MUST support all characters defined in the "OpaqueString" profile defined in Section 4.2 of [RFC7613].

- Constructs the user-pass by concatenating the user-id, a single colon (":") character, and the password

Basically in the token what is before the first ':' is to be considered username what is before is to be considered password, and both user and password con be non alphanumeric and not even ASCII

Reply:
The line is now `if(clear_token[i] == ':') { colonIndex = i; break; }`